test/engine.spec.js

```diff
@@ -286,4 +354,82 @@ describe('basic tests', function () {
       done()
     })
   })
+
+  it('handles error in next handler', function (done) {
+    const engine = new RpcEngine()
+
+    engine.push(function (_req, _res, next, _end) {
+      next(function (_cb) {
+        throw new Error('foo')
+      })
+    })
+
+    engine.push(function (_req, res, _next, end) {
+      res.result = 42
+      end()
+    })
+
+    const payload = { id: 1, jsonrpc: '2.0', method: 'hello' }
+
+    engine.handle(payload, (err, _res) => {
+      assert.ok(err, 'did error')
+      assert.equal(err.message, 'foo', 'error has expected message')
+      done()
+    })
+  })
+
+  it('handles failure to end request', function (done) {
+    const engine = new RpcEngine()
+
+    engine.push(function (_req, res, next, _end) {
+      res.result = 42
+      next()
+    })
+
+    const payload = { id: 1, jsonrpc: '2.0', method: 'hello' }
+
+    engine.handle(payload, (err, res) => {
+      assert.ok(err, 'should have errored')
+      assert.ok(
+        err.message.startsWith('JsonRpcEngine: Nothing ended request:'),
+        'should have expected error message',
+      )
+      assert.ok(!res.result, 'should not have result')
+      done()
+    })
+  })
+
+  it('handles batch request processing error', function (done) {
+    const origPromiseAll = Promise.all
+    Promise.all = () => {
+      Promise.all = origPromiseAll
+      throw new Error('foo')
+    }
+
+    const engine = new RpcEngine()
```

Comment:
Woah, mutating `Promise`? That seems a bit excessive. That could lead to some pretty nasty test failures if it was left unrestored somehow. I think you could stub out `_promiseHandle` here instead, so that only the instance under test is mutated.

e.g.
```suggestion
    const engine = new RpcEngine()
    sinon.stub(engine, '_promiseHandle').throws('foo')
```

The same goes for the following test as well

Reply:
Look man, we're all professionals here 😅 

dab68d5

Reply:
That's: 62d731d